deploy/ciao-deploy-client/main.go

```diff
@@ -0,0 +1,474 @@
+	systemCertPath := path.Join(systemPkiLocation, deploy.CertName(ssntpRole))
+
+	// TODO: Check if certificate installed is valid
+	if !force {
+		if _, err := os.Stat(systemCertPath); !os.IsNotExist(err) {
+			fmt.Printf("Certificate already installed. Skipping download.\n")
+			return systemCertPath, nil
+		}
+	}
+
+	fp, err := getCSRFingerprint()
+	if err != nil {
+		return "", errors.Wrap(err, "error getting CSR fingerprint")
+	}
+
+	err = downloadSignedCert(systemCertPath, fp)
+	if err != nil {
+		return "", errors.Wrap(err, "error downloading signed certificate")
+	}
+
+	// Cleanup saved CSR and private key
+	os.Remove(privPath(ssntpRole))
+	os.Remove(csrPath(ssntpRole))
+
+	return systemCertPath, nil
+}
+
+// Provides path to CA certificate. If not locally available then this is downloaded from the server.
+func obtainCACert() (string, error) {
+	systemCaCertPath := path.Join(systemPkiLocation, "CAcert.pem")
+
+	// TODO: Check if certificate installed is valid
+	if !force {
+		if _, err := os.Stat(systemCaCertPath); !os.IsNotExist(err) {
+			fmt.Printf("CA certificate already installed. Skipping download.\n")
+			return systemCaCertPath, nil
+		}
+	}
+
+	url := fmt.Sprintf("%s/cacert", deployServerHost)
+	fmt.Printf("Downloading CA certificate: %s\n", url)
+	r, err := http.Get(url)
+	if err != nil {
+		return "", errors.Wrap(err, "Error downloading certificate")
+	}
+	defer r.Body.Close()
+
+	if r.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("Expected OK when downloading certificate: %s", r.Status)
+	}
+
+	var certOutput bytes.Buffer
+	io.Copy(&certOutput, r.Body)
+	certBlock, _ := pem.Decode(certOutput.Bytes())
+	cert, err := x509.ParseCertificate(certBlock.Bytes)
+	if err != nil {
+		return "", errors.Wrap(err, "Error parsing CA certificate")
+	}
+
+	fp := certs.FingerPrint(cert)
+	fmt.Printf("Got CA cert with fingerprint: %s\n", fp)
+
+	f, err := ioutil.TempFile("", "CAcert.pem")
+	if err != nil {
+		return "", errors.Wrap(err, "Error creating CA certificate file")
+	}
+	defer os.Remove(f.Name())
+
+	_, err = io.Copy(f, &certOutput)
+	if err != nil {
+		return "", errors.Wrap(err, "Error copying CA certificate to file")
+	}
+
+	if err := f.Chmod(0644); err != nil {
+		return "", errors.Wrap(err, "Error changing mode on certificate")
+	}
+	f.Close()
+
+	if err := deploy.SudoCopyFile(systemCaCertPath, f.Name()); err != nil {
+		return "", errors.Wrap(err, "Error copying CA certificate to final destination")
+	}
+
+	fmt.Printf("CA certificate installed in: %s\n", systemCaCertPath)
+	return systemCaCertPath, nil
+}
+
+func refreshClient() error {
+	if !downloadClient {
+		return nil
+	}
+
+	downloadedClientPath, err := deploy.DownloadTool(deployServerHost, "ciao-deploy-client")
+	if err != nil {
+		errors.Wrap(err, "Error when downloading client")
+	}
+	defer os.Remove(downloadedClientPath)
+
+	fmt.Println("Downloaded client")
+
+	if _, err := os.Stat(downloadedClientPath); !os.IsNotExist(err) {
```

Comment:
I don't really get what this if statement is checking.  Presumably, the file exists if DownloadTool returns no error, although I haven't looked at the code of this function yet.   Also, I don't really understand what the conditions for entering this block our.  Currently, we come in here, if stat succeeded or if stat failed with some reason other than does not exist.  Is this what we want?